src/psyclone/transformations.py

```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+                                                                  node)
+            if forward_dep is None:
+                continue
+            # If the forward dependence is always in the same Parallel
+            # Region or the parent Parallel region
+            # Check if the taskloop and its forward dependence are in the
+            # same serial region
+            if (taskloop.ancestor(OMPSerialDirective) is not
+                    forward_dep.ancestor(OMPSerialDirective)):
+                # They're not in the same serial region. Check if our
+                # taskloop is in a waiting Single Directive
+                valid = True
+                ancestor = taskloop.ancestor(OMPSerialDirective)
+                valid = valid and isinstance(ancestor,
+                                             nodes.OMPSingleDirective)
+                if valid:
+                    valid = valid and (not ancestor.nowait)
+                # If not valid, then we're in a Master Directive or a
+                # Single Directive with nowait. In this case we can't
+                # safely guarantee our forward dependency so throw an
+                # error
+                if not valid:
+                    raise TransformationError("Couldn't satisfy the "
+                                              "dependencies due to "
+                                              "taskloop dependencies "
+                                              "across barrierless OMP "
+                                              "serial regions.")
+
+    @staticmethod
+    def get_forward_dependence(taskloop, root):
+        '''
+        Returns the next forward dependence for a taskloop using the new
+        dependency analysis functions.
+        Forward dependencies can be of the following types:
+        Loop
+        OMPDoDirective
+        OMPTaskloopDirective
+        OMPTaskwaitDirective (If in same OMPSingle and that single has
+        nowait=False)
+        OMPSingleDirective (If ancestor OMPSingle has nowait=False)
+        OMPParallelDirective
+
+        The forward dependency is never a child of taskloop, and must have
+        abs_position > taskloop.abs_position
+
+        :param taskloop: the taskloop node to search for the
+                         forward_dependence.
+        :type taskloop: :py:class:`psyclone.psyir.nodes.OMPTaskloopDirective
+        :param root: the tree in which to search for the forward_dependence.
+        :type root: :py:class:`psyclone.f2pygen.DoGen`
+
+        :returns: the forward_dependence of taskloop.
+        :rtype: :py:class:`psyclone.f2pygen.DoGen`
+        '''
+        # We only look for specific types
+        node_list = root.walk((Loop, OMPDoDirective, OMPTaskloopDirective,
+                               OMPTaskwaitDirective, OMPSingleDirective,
+                               OMPParallelDirective))
+        taskloop_vars = VariablesAccessInfo()
+        for child in taskloop.walk(nodes.Node):
```

Comment:
I will add a comment explaining this. Essentially the functionality that I need is not quite provided in `dependency_tools`. The `VariablesAccessInfo` that results from using the dependency_tools directly is the read/write accesses to the loop variable (by the loops), which means that since loop variables are probably reused by each loop, all loops point dependencies to each other. The easiest way I could think of to do this is to implement it myself, where I skip over all Loops explicitly.